daemon/logger/jsonfilelog/read.go

```diff
@@ -153,42 +174,42 @@ func followLogs(f *os.File, logWatcher *logger.LogWatcher, notifyRotate chan int
 					retries++
 					continue
 				}
-				logWatcher.Err <- err
+
 				return
 			}
 
-			logrus.WithField("logger", "json-file").Debugf("waiting for events")
-			if err := fileWatcher.Add(f.Name()); err != nil {
-				logrus.WithField("logger", "json-file").Warn("falling back to file poller")
-				fileWatcher.Close()
-				fileWatcher = filenotify.NewPollingWatcher()
-				if err := fileWatcher.Add(f.Name()); err != nil {
-					logrus.Errorf("error watching log file for modifications: %v", err)
-					logWatcher.Err <- err
-				}
-			}
 			select {
 			case <-fileWatcher.Events():
 				dec = json.NewDecoder(f)
-				fileWatcher.Remove(f.Name())
 				continue
 			case <-fileWatcher.Errors():
-				fileWatcher.Remove(f.Name())
 				logWatcher.Err <- err
 				return
 			case <-logWatcher.WatchClose():
-				fileWatcher.Remove(f.Name())
+				fileWatcher.Remove(name)
 				return
 			case <-notifyRotate:
-				f, err = os.Open(f.Name())
+				f.Close()
+				fileWatcher.Remove(name)
+
+				// retry when the file doesn't exist
+				for retries := 0; retries <= 5; retries++ {
```

Comment:
It is, but not with the notification itself, but rather that the file can be rotated again by the time we're handling the notification. In most cases this won't be hit, but in my testing I did hit it.